serverless/internal/git/git.go

```diff
@@ -0,0 +1,253 @@
+
+// git contains libraries for using github repositories that make serverless
+// operations easy to follow.
+package git
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-billy/v5/util"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/golang/glog"
+	"github.com/google/go-github/v37/github"
+	"golang.org/x/oauth2"
+)
+
+// RepoID identifies a github project, including owner &repo.
+type RepoID struct {
+	Owner    string
+	RepoName string
+}
+
+// String returns "<owner>/<repo>".
+func (r RepoID) String() string {
+	return fmt.Sprintf("%s/%s", r.Owner, r.RepoName)
+}
+
+// NewRepoID creates a new RepoID struct from an owner/repo fragment.
+func NewRepoID(or string) (RepoID, error) {
+	s := strings.Split(or, "/")
+	if l, want := len(s), 2; l != want {
+		return RepoID{}, fmt.Errorf("can't split owner/repo %q, found %d parts want %d", or, l, want)
+	}
+	return RepoID{
+		Owner:    s[0],
+		RepoName: s[1],
+	}, nil
+}
+
+// NewServerlessRepo creates a wrapper around a git repository which has a fork owned by
+// the user, and an upstream repository configured that PRs can be proposed against.
+func NewServerlessRepo(ctx context.Context, upstream, fork RepoID, ghUser, ghEmail, ghToken, clonePath string) (ForkedRepo, error) {
+	repo := ForkedRepo{
+		upstream: upstream,
+		fork:     fork,
+		user:     ghUser,
+		email:    ghEmail,
+	}
+	var err error
+	repo.ghCli, err = authWithGithub(ctx, ghToken)
+	if err != nil {
+		return repo, err
+	}
+
+	// git clone -o origin "https://${GIT_USERNAME}:${FEEDER_GITHUB_TOKEN}@github.com/${fork_repo}.git" "${temp}"
+	cloneOpts := &git.CloneOptions{
+		URL: fmt.Sprintf("https://%v:%v@github.com/%s.git", ghUser, ghToken, repo.fork),
+	}
+	if clonePath == "" {
+		repo.git, err = git.Clone(memory.NewStorage(), memfs.New(), cloneOpts)
+	} else {
+		repo.git, err = git.PlainClone(clonePath, false, cloneOpts)
+	}
+	if err != nil {
+		return repo, fmt.Errorf("failed to clone fork repo %q: %v", cloneOpts.URL, err)
+	}
+
+	// Create a remote -> upstreamRepo
+	//  git remote add upstream "https://github.com/${upstream_repo}.git"
+	upstreamURL := fmt.Sprintf("https://github.com/%v.git", upstream)
+	upstreamRemote, err := repo.git.CreateRemote(&config.RemoteConfig{
+		Name: "upstream",
+		URLs: []string{upstreamURL},
+	})
+	if err != nil {
+		return repo, fmt.Errorf("failed to add remote %q to fork repo: %v", upstreamURL, err)
+	}
+	glog.V(1).Infof("Added remote upstream->%q for %q:\n%v", upstreamURL, repo.fork, upstreamRemote)
+
+	if err = repo.git.FetchContext(ctx, &git.FetchOptions{}); err != nil && err != git.NoErrAlreadyUpToDate {
+		return repo, fmt.Errorf("failed to fetch repo %q: %v", repo.fork, err)
+	}
+	return repo, err
```

Comment:
nit (on me): I think this should probably be `return repo, nil` (just in case err is `git.NoErrAlreadyUpToDate`)?

Reply:
Done.